### PR TITLE
Use f64::total_cmp instead of OrderedFloat

### LINKED ISSUE #4051: Use f64::total_cmp instead of OrderedFloat
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**

DataFusion uses OrderedFloat in various places to order floats, this is different from arrow-rs which makes use of https://doc.rust-lang.org/std/primitive.f32.html#method.total_cmp.

**Describe the solution you'd like**

Moving away from OrderedFloat will:

* Ensure consistency
* Eliminate a dependency

**Describe alternatives you've considered**

**Additional context**

### PR DESCRIPTION
**Pull request by comphead**
# Which issue does this PR close?

Closes #4051 .

# Rationale for this change
See #4051 

# What changes are included in this PR?

# Are there any user-facing changes?

### CODE REVIEW

**Comment by tustvold** [comment]
@comphead I would recommend creating a newtype wrapper around a num::Float that implements Hash using hash_utils, eq using total_cmp, etc...

**Comment by comphead** [comment]
> @comphead I would recommend creating a newtype wrapper around a float that implements Hash using hash_utils, eq using total_cmp, etc...

Hi @tustvold I have implemented hasher through `std::hash` but the impl is the same as in hash_utils. HashValue trait is not the same as Hash, afaik. Let me know if the hasher should be done in other way

File: datafusion/common/src/scalar.rs [hunk 1]
```diff
@@ -221,17 +221,9 @@ impl PartialOrd for ScalarValue {
             (Decimal128(_, _, _), _) => None,
             (Boolean(v1), Boolean(v2)) => v1.partial_cmp(v2),
             (Boolean(_), _) => None,
-            (Float32(v1), Float32(v2)) => {
-                let v1 = v1.map(OrderedFloat);
-                let v2 = v2.map(OrderedFloat);
-                v1.partial_cmp(&v2)
-            }
+            (Float32(v1), Float32(v2)) => v1.partial_cmp(v2),
```
Comment (tustvold) [inline datafusion/common/src/scalar.rs:224]:
```suggestion
            (Float32(v1), Float32(v2)) => v1.total_cmp(v2),
```

File: datafusion/common/src/scalar.rs [hunk 2]
```diff
@@ -637,14 +643,8 @@ impl std::hash::Hash for ScalarValue {
...
             Boolean(v) => v.hash(state),
-            Float32(v) => {
-                let v = v.map(OrderedFloat);
-                v.hash(state)
-            }
-            Float64(v) => {
-                let v = v.map(OrderedFloat);
-                v.hash(state)
-            }
+            Float32(v) => v.map(Fl).hash(state),
```
Comment (tustvold) [inline datafusion/common/src/scalar.rs:646]: I think this can just call HashValue on v?

File: datafusion/common/src/scalar.rs [hunk 3]
```diff
@@ -132,15 +129,19 @@ impl PartialEq for ScalarValue {
             (Boolean(v1), Boolean(v2)) => v1.eq(v2),
             (Boolean(_), _) => false,
             (Float32(v1), Float32(v2)) => {
-                let v1 = v1.map(OrderedFloat);
-                let v2 = v2.map(OrderedFloat);
-                v1.eq(&v2)
+                // Handle NaN == NaN as true manually like in OrderedFloat
```
Comment (tustvold) [inline datafusion/common/src/scalar.rs:132]: To be consistent with the hash implementation, this should also use total_cmp. Otherwise two equal values, e.g. +0 and -0, will have different hashes

File: datafusion/physical-expr/src/aggregate/tdigest.rs [hunk 4]
```diff
@@ -788,10 +728,7 @@ mod tests {
 
         for _ in 1..=100 {
             let t = TDigest::new(100);
-            let values: Vec<_> = (1..=1_000)
-                .map(f64::from)
-                .map(|v| OrderedFloat::from(v as f64))
-                .collect();
+            let values: Vec<_> = (1..=1_000).map(f64::from).map(|v| v as f64).collect();
```
Comment (tustvold) [inline datafusion/physical-expr/src/aggregate/tdigest.rs:731]:
```suggestion
            let values: Vec<_> = (1..=1_000).map(f64::from).collect();
```

File: datafusion/physical-expr/src/aggregate/tdigest.rs [hunk 5]
```diff
@@ -766,13 +711,8 @@ mod tests {
...
     fn test_merge_unsorted_against_skewed_distro() {
         let t = TDigest::new(100);
-        let mut values: Vec<_> = (1..=600_000)
-            .map(f64::from)
-            .map(|v| OrderedFloat::from(v as f64))
-            .collect();
-        for _ in 0..400_000 {
-            values.push(OrderedFloat::from(1_000_000_f64));
-        }
+        let mut values: Vec<_> = (1..=600_000).map(f64::from).map(|v| v as f64).collect();
```
Comment (tustvold) [inline datafusion/physical-expr/src/aggregate/tdigest.rs:714]:
```suggestion
        let mut values: Vec<_> = (1..=600_000).map(f64::from).collect();
```

File: datafusion/physical-expr/src/aggregate/tdigest.rs [hunk 6]
```diff
@@ -748,10 +696,7 @@ mod tests {
     #[test]
     fn test_merge_unsorted_against_uniform_distro() {
         let t = TDigest::new(100);
-        let values: Vec<_> = (1..=1_000_000)
-            .map(f64::from)
-            .map(|v| OrderedFloat::from(v as f64))
-            .collect();
+        let values: Vec<_> = (1..=1_000_000).map(f64::from).map(|v| v as f64).collect();
```
Comment (tustvold) [inline datafusion/physical-expr/src/aggregate/tdigest.rs:699]:
```suggestion
        let values: Vec<_> = (1..=1_000_000).map(f64::from).collect();
```

File: datafusion/physical-expr/src/aggregate/tdigest.rs [hunk 7]
```diff
@@ -658,7 +608,7 @@ impl TDigest {
 
         let max = cast_scalar_f64!(&state[3]);
         let min = cast_scalar_f64!(&state[4]);
-        assert!(max >= min);
+        assert!(max.is_nan() || min.is_nan() || max >= min);
```
Comment (tustvold) [inline datafusion/physical-expr/src/aggregate/tdigest.rs:611]: Why this change?

File: datafusion/common/src/scalar.rs [hunk 8]
```diff
@@ -132,15 +129,19 @@ impl PartialEq for ScalarValue {
...
             (Float32(_), _) => false,
             (Float64(v1), Float64(v2)) => {
-                let v1 = v1.map(OrderedFloat);
-                let v2 = v2.map(OrderedFloat);
-                v1.eq(&v2)
+                // Handle NaN == NaN as true manually like in OrderedFloat
+                match (v1, v2) {
+                    (Some(f1), Some(f2)) if f1.is_nan() && f2.is_nan() => true,
+                    _ => v1.eq(v2),
+                }
```
Comment (tustvold) [inline datafusion/common/src/scalar.rs:144]:
```suggestion
               v1.to_bits() == v2.to_bits()
```

File: datafusion/common/src/scalar.rs [hunk 9]
```diff
@@ -132,15 +129,19 @@ impl PartialEq for ScalarValue {
...
             (Boolean(_), _) => false,
             (Float32(v1), Float32(v2)) => {
-                let v1 = v1.map(OrderedFloat);
-                let v2 = v2.map(OrderedFloat);
-                v1.eq(&v2)
+                // Handle NaN == NaN as true manually like in OrderedFloat
+                match (v1, v2) {
+                    (Some(f1), Some(f2)) if f1.is_nan() && f2.is_nan() => true,
+                    _ => v1.eq(v2),
+                }
```
Comment (tustvold) [inline datafusion/common/src/scalar.rs:136]:
```suggestion
                v1.to_bits() == v2.to_bits()
```

File: datafusion/physical-expr/src/aggregate/tdigest.rs (hunk 4 above)
Comment (comphead) [reply datafusion/physical-expr/src/aggregate/tdigest.rs:731]: Done

File: datafusion/physical-expr/src/aggregate/tdigest.rs (hunk 5 above)
Comment (comphead) [reply datafusion/physical-expr/src/aggregate/tdigest.rs:714]: Done

File: datafusion/physical-expr/src/aggregate/tdigest.rs (hunk 6 above)
Comment (comphead) [reply datafusion/physical-expr/src/aggregate/tdigest.rs:699]: Done

File: datafusion/common/src/scalar.rs (hunk 1 above)
Comment (comphead) [reply datafusion/common/src/scalar.rs:224]: v1 is an `Option<f32>`, it supports partial_cmp, not total_cmp. let me know if you ok if I unwrap it to floats, the same way as done for Decimals.

File: datafusion/common/src/scalar.rs (hunk 3 above)
Comment (comphead) [reply datafusion/common/src/scalar.rs:132]: What if 
```
                match (v1, v2) {
                    (Some(f1), Some(f2)) => f1.total_cmp(f2).is_eq(),
                    _ => v1.eq(v2),
                }
```

File: datafusion/common/src/scalar.rs (hunk 8 above)
Comment (tustvold) [reply datafusion/common/src/scalar.rs:144]: Sounds good, I would recommend using `to_bits` instead of `total_tmp().is_eq()`, as LLVM isn't smart enough to figure out that the latter is just bitwise equality - https://github.com/apache/arrow-rs/pull/3060

File: datafusion/common/src/scalar.rs (hunk 1 above)
Comment (tustvold) [reply datafusion/common/src/scalar.rs:224]: Yes, we will need to unwrap the option, I keep forgetting that ScalarValue has typed nulls for some reason :laughing: 

partial_cmp on Option, will call through to partial_cmp on f32, which is not the same as total_cmp

File: datafusion/physical-expr/src/aggregate/tdigest.rs (hunk 7 above)
Comment (comphead) [reply datafusion/physical-expr/src/aggregate/tdigest.rs:611]: max/min can be none, direct f64 comparison works not correctly in this case. replaced with `assert(max.total_cmp(&min).is_ge());`

File: datafusion/common/src/scalar.rs (hunk 9 above)
Comment (comphead) [reply datafusion/common/src/scalar.rs]: v1 is `Option<f32>`
what if
```
                match (v1, v2) {
                    (Some(f1), Some(f2)) => f1.total_cmp(f2).is_eq(),
                    _ => v1.eq(v2),
                }
```

File: datafusion/common/src/scalar.rs (hunk 2 above)
Comment (tustvold) [reply datafusion/common/src/scalar.rs:646]: Fair, I think there is a way to clean this up, but we can do that in a follow on PR

File: datafusion/common/src/scalar.rs (hunk 1 above)
Comment (comphead) [reply datafusion/common/src/scalar.rs]: Done. Yeah, I checked that Float64(NULL) == Float64(NULL) now.

File: datafusion/common/src/scalar.rs [hunk 10]
```diff
@@ -221,17 +215,9 @@ impl PartialOrd for ScalarValue {
             (Decimal128(_, _, _), _) => None,
             (Boolean(v1), Boolean(v2)) => v1.partial_cmp(v2),
             (Boolean(_), _) => None,
-            (Float32(v1), Float32(v2)) => {
-                let v1 = v1.map(OrderedFloat);
-                let v2 = v2.map(OrderedFloat);
-                v1.partial_cmp(&v2)
-            }
+            (Float32(Some(f1)), Float32(Some(f2))) => Some(f1.total_cmp(f2)),
```
Comment (tustvold) [inline datafusion/common/src/scalar.rs:218]: I think this will now return None when comparing nulls, which isn't consistent with the other types

Comment (comphead) [reply datafusion/common/src/scalar.rs]: Right! Fixed.

**Review by tustvold — APPROVED** [review]
Thank you :+1:
